cmd/ci-operator-configresolver/main.go

```diff
@@ -0,0 +1,251 @@
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"net/http"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	"github.com/openshift/ci-tools/pkg/config"
+	"github.com/openshift/ci-tools/pkg/load"
+	"github.com/openshift/ci-tools/pkg/registry"
+	types "github.com/openshift/ci-tools/pkg/steps/types"
+)
+
+type options struct {
+	resolver   registry.Resolver
+	configPath string
+	regPath    string
+}
+
+type coalescer struct {
+	sync.Mutex
+	loading bool
+	loader  *sync.Cond
+	latest  *registry.Resolver
+}
+
+func (c *coalescer) reload(o options) {
+	c.Lock()
+	if c.loading {
+		// someone else is loading it, so we wait
+		log.Debug("Registry already being reloaded; waiting")
+		c.loader.L.Lock()
+		c.Unlock()
+		c.loader.Wait()
+		c.loader.L.Unlock()
+		log.Debug("Registry reloaded by separate goroutine, returning")
+		return
```

Comment:
How does this work? I don't see you return anything and I don't see anything use the `latest` value. Whatever is going to be accessing `latest` will also be racy unless it locks, so IMO you should return a reference here since all the locking is already handled.

Reply:
(unit tests help)

Reply:
Make sure to run `go test -race` to catch data races